### PR TITLE
Sandbox: Fix `history.replace` calls inside the sandbox

### PR DESCRIPTION
**Pull request by academo**
**What is this feature?**

This function adds a new "live" distortion for the `history` object. specifically the `history.replace` function.

### The problem

When a plugin calls `history.replace(object)` that `object` is a proxy. `history` will later take that object and pass it to
`window.history.replaceState` which will fail because it can't handle proxies (of any kind).

To prevent that, we need to either monkey-patch `window.history.replaceState` or distort `history.replace` to prevent it from calling `window.history.replaceState` with a proxy.

**Why do we need this feature?**

To make plugins using history.replace work correctly inside the sandbox

**Who is this feature for?**

plugins running inside the sandbox

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**

Please check that:

- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Review by jackw — APPROVED** [review]
LGTM! 🚀

File: public/app/features/plugins/sandbox/distortion_map.ts [hunk 1]
```diff
@@ -520,13 +521,30 @@ async function distortPostMessage(distortions: DistortionMap) {
...
+ * or because the libraries we want to patch are lazy-loaded and we don't have access to their definitions.
+ * We put here only distortions that can't be static because they are dynamicly loaded
  */
-export async function distortLiveApis() {
+export function distortLiveApis(originalValue: ProxyTarget): ProxyTarget | undefined {
   distortMonacoEditor(generalDistortionMap);
+
+  // This distorts the `history.replace` function in react-router-dom.
+  // constructed for each browser history and is only accessible within the react context.
+  // Note that this distortion does not affect `String.prototype.replace` calls.
```
Comment (leventebalogh) [inline public/app/features/plugins/sandbox/distortion_map.ts:534]: What is preventing the distortion below to jump on the `String.prototype.replace` calls? Is it the `originalValue.prototype.constructor.length === 2` part?

**Review by leventebalogh — APPROVED** [review]
Looks good to me, left one question. 👍

File: public/app/features/plugins/sandbox/distortion_map.ts (hunk 1 above)
Comment (academo) [reply public/app/features/plugins/sandbox/distortion_map.ts:534]: Calls to Js intrinsic apis such as string don't go through distortions. Let me add that in the comment to make it clearer

**Pull request merged**